### PR TITLE
qmatmul for i8xu8 and friends

### LINKED ISSUE #413: i8xu8 (and friends) matrix products
Can we just use the i8xi8 path by applying an offset and playing with zeropoint ?

### PR DESCRIPTION
**Pull request by janpetschexain**
**Problem**

fixes #413 

**Solution**

`tract-data`:
- add `Tensor::offset_u8_as_i8()` to apply the offsetting to a tensor elementwise

`tract-core`:
- add `ops::quant::OffsetU8asI8` operation to be wired in the quantized matmul
- conditionally wire the offsettings in `QMatMul` and `QMatMulUnary` in the evaluation and decluttering/codegen
- generalize the existing quantized matmul tests with a macro and add new test cases for the various i8-u8 combinations

**Open Questions**

- how to handle the `bias` term is a bit unclear to me. it is always of type `i32` and the way it is used in `wire_matmul_quant()` means that it is only affected by the scales but not the zero points (i would assume that the bias is expected to be given in a scaled form, but this would affect the computation of the scales which i didn't observe, so maybe it is compensated via the `c_scale`). conversely, we probably also don't have to apply `wire_offset_u8_as_i8()` to it, which would also have no effect on it in its current implementation. unfortunatetly, i didn't find any examples in the code except for `None`/`0` `bias`.

### CODE REVIEW

File: core/src/ops/matmul/mir_quant.rs [hunk 1]
```diff
@@ -731,149 +802,330 @@ mod test {
...
-        fn check(&self) {
-            let r = self.reference();
-            let t = self.tract();
-            if r.iter().zip(t.iter()).any(|(r, t)| r.max(t) - r.min(t) > 1) {
-                panic!("mismatch! refernce: {:?} tract: {:?}", r, t)
+    fn scale() -> BoxedStrategy<f32> {
+        prop_oneof![Just(1.0), (1i32..=20).prop_map(|x| x as f32 / 20.0)].boxed()
+    }
+
+    macro_rules! impl_qmmp {
```
Comment (kali) [inline core/src/ops/matmul/mir_quant.rs:816]: Mostly curious... can we not achieve this with type parameters on QProblem instead of a macro ?

File: core/src/ops/matmul/mir_quant_unary.rs [hunk 2]
```diff
@@ -535,135 +583,309 @@ mod test {
...
-        fn check(&self) {
-            let r = self.reference();
-            let t = self.tract();
-            if r.iter().zip(t.iter()).any(|(r, t)| r.max(t) - r.min(t) > 1) {
-                panic!("mismatch! refernce: {:?} tract: {:?}", r, t)
+    fn scale() -> BoxedStrategy<f32> {
+        prop_oneof![Just(1.0), (1i32..=20).prop_map(|x| x as f32 / 20.0)].boxed()
+    }
+
+    macro_rules! impl_qmmup {
```
Comment (kali) [inline core/src/ops/matmul/mir_quant_unary.rs:597]: same question here.

**Review by kali — APPROVED** [review]

**Comment by kali** [comment]
Thanks, this looks like a very high-quality contribution. 

Would you mind reactivating the "full test harness" workflow or run it manually in your clone ? Or please confirm ./.travis/native.sh runs on your system :)

We will quickly have a look at the bias question, I would like this to be merged quickly.

**Comment by kali** [comment]
We discussed the bias question here. Basically, the consensus is that we need the bias here because it has to be added to the accumulators before they are quantized (or saturation would interfere). So the assumption is that the bias is "ready" to be consumed and added to the accumulators, in i32, in a scale consistent with the product of input scales, and without bias.

We have only one real-life example that make use of bias in Q nets and it's on the side, not in this repository already. It does work under this assumption. It's a quantized NNEF inception network that we will had to the test harness and/or examples soon.

So basically... Unless you have other real-life examples that will not work like this, I think we should let it be.

File: core/src/ops/matmul/mir_quant.rs (hunk 1 above)
Comment (janpetschexain) [reply core/src/ops/matmul/mir_quant.rs:816]: i tried that in the beginning and was running into trait bound issues, iirc not so much for `Datum` but more for `Arbitrary`. it might be possible to express this without a macro now in the latest iteration of the tests, but i haven't tried that.

**Comment by janpetschexain** [comment]
> Would you mind reactivating the "full test harness" workflow or run it manually in your clone ? Or please confirm ./.travis/native.sh runs on your system :)

running the `native.sh` script finally finished, everything passes successfully, this is the output on my machine:

<details>
<summary>native.sh output</summary>
+ ./.travis/regular-tests.sh
+ cargo -q test -q -p tract-core -p tract-hir -p tract-onnx -p tract-linalg

running 202 tests
.................................................................................................... 100/202
.................................................................................................... 200/202
..
test result: ok. 202 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 60.48s


running 76 tests
............................................................................
test result: ok. 76 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s


running 1103 tests
.................................................................................................... 100/1103
.................................................................................................... 200/1103
.................................................................................................... 300/1103
.................................................................................................... 400/1103
.................................................................................................... 500/1103
.................................................................................................... 600/1103
.................................................................................................... 700/1103
.................................................................................................... 800/1103
.................................................................................................... 900/1103
.................................................................................................... 1000/1103
.................................................................................................... 1100/1103
...
test result: ok. 1103 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 35.12s


running 2 tests
..
test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 1 test
test src/lib.rs - (line 7) ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.30s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ cargo -q test -q -p tract-tensorflow --lib

running 36 tests
....................................
test result: ok. 36 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

+ '[' -n '' ']'
+ cargo -q test -q --release -p core-proptest-pulse

running 34 tests
i.................................
test result: ok. 33 passed; 0 failed; 1 ignored; 0 measured; 0 filtered out; finished in 0.49s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ cargo -q test -q --release -p lstm-proptest-onnx-vs-tf

running 6 tests
......
test result: ok. 6 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.43s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ cargo -q test -q --release -p nnef-inceptionv3

running 1 test
.
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 15.14s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ cargo -q test -q --release -p tf-inceptionv3

running 1 test
.
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 10.04s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ cargo -q test -q --release -p tf-mobilenet-v2

running 3 tests
...
test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 5.38s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ cargo -q test -q --release -p tf-moz-deepspeech

running 3 tests
iii
test result: ok. 0 passed; 0 failed; 3 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ cargo -q test -q -p onnx-test-suite -- --skip real_

running 1 test
.
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 2847 tests
iiii..................................................i..i.i..i.......i.i........i....i...i..i...... 100/2847
......iiii..............iiii....i...i..........iiiiiiiiiiiiiii.....ii.i.iiiiii.iiiii..........iii... 200/2847
..........i...........i................i.............i..iiii...........................i..ii..i..... 300/2847
...i.............i.i....i....iiii....i............iiiiiiiiiiiiiii...ii...ii.i.i.i.i.i.i.i.i.i.i.i.i. 400/2847
i.iii.i.i.iiiiiiiiii...............................i..iiiiii........................................ 500/2847
...........................................i.........iiiiiiiiiiiiiiiiiiiiiiiii.i..i......iiiiiiiiiii 600/2847
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii...ii.......i..............i............i.. 700/2847
..........i....iiiiii...........iiiiiii.....iiiiiiiii.......iiiiiiiiiiiiiiiiiiiiiii........i........ 800/2847
.......iiii....................................................................i.i........i....i.... 900/2847
..i............iiii..............iiii....i...i..........iiiiiiiiiii.iii.....ii.i..i.i.i.iiiii....... 1000/2847
...................i...........i................i.............i..iiii...........................i..i 1100/2847
i..i........i...............i..........iii....i.........................ii...ii...ii.i.i.i.i.i.i.i.i 1200/2847
.i.i.i.i.i.i.i.i.i.iiiiiiiii................................i...iiiii............................... 1300/2847
....................................................i.........iiiiiiiiiiiiiiii.iiiiiiii.i..i......ii 1400/2847
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii...ii.......i..............i...... 1500/2847
......i............i....iiiiii...........iiiiiii.....iiiiiiiii.......iiiiiiiiiiiiiiiiiiiiiii........ 1600/2847
i...............iiii....................................................................i.i........i 1700/2847
....i......i..........................................i..........iiiiiiiiiii.iii......i.i..i.....iii 1800/2847
ii..........................i...........i................i.............i..iiii...................... 1900/2847
.....i..ii..i........i...............i...........iii...i.........................ii...ii...ii.i.i.i. 2000/2847
i.i.i.i.i.i.i.i.i.i.i.i.i.i.iiiiiiiii....................................iiiii...................... 2100/2847
.............................................................i.........iiiiiiiiiiiiiiii.iiiiiiii.i.. 2200/2847
i......iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii...ii.......i............ 2300/2847
...............i.................iiiiii...........iiiiiii.....iiiiiiiii.......iiiiiiiiiiiiiiiiiiiiii 2400/2847
i........i...............................................................i.i....i.i.i............... 2500/2847
.......................................................i.i....i.i.i................................. 2600/2847
.....................................i.i....i.i.i.................................................i. 2700/2847
.................................i..................................i......iiiiiiiiii...iiiiii....ii 2800/2847
iiiiiiii...iiiiii....iiiiiiiiii...iiiiii.......
test result: ok. 2023 passed; 0 failed; 824 ignored; 0 measured; 27 filtered out; finished in 344.43s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ ./.travis/onnx-tests.sh
+ opset=onnx_1_9_0
+ cd harness/onnx-test-suite
+ cargo -q check -q --no-default-features --features onnx_1_9_0
+ cargo -q test -q --no-default-features --release --features onnx_1_9_0

running 1 test
.
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s


running 2874 tests
iiii..................................................i...ii..i.......i.i........i....i...i..i...... 100/2874
......iiii..............iiii....i...i..........iiiiiiiiiiiiiii.....ii.i.iiiiii.iiiii..........iii... 200/2874
..........i...........i................i.............i..iiii...........................i..ii..i..... 300/2874
...i.............i.i....i....iiii....i............iiiiiiiiiiiiiii...ii...ii.i.i.i.i.i.i.i.i.i.i.i.i. 400/2874
i.iii.i.i.iiiiiiiiii...............................i..iiiiii........................................ 500/2874
...........................................i.........iiiiiiiiiiiiiiiiiiiiiiiii.i..i......iiiiiiiiiii 600/2874
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii...ii.......i..............i............i.. 700/2874
..........i....iiiiii...........iiiiiii.....iiiiiiiii.......iiiiiiiiiiiiiiiiiiiiiii........i........ 800/2874
.......iiii....................................................................i.i........i....i.... 900/2874
..i............iiii..............iiii....i...i..........iiiiiiiiiii.iii.....ii.i..i.i.i.iiiii....... 1000/2874
...................i...........i................i.............i..iiii...........................i..i 1100/2874
i..i........i...............i..........iii....i.........................ii...ii...ii.i.i.i.i.i.i.i.i 1200/2874
.i.i.i.i.i.i.i.i.i.iiiiiiiii................................i...iiiii............................... 1300/2874
....................................................i.........iiiiiiiiiiiiiiii.iiiiiiii.i..i......ii 1400/2874
iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii...ii.......i..............i...... 1500/2874
......i............i....iiiiii...........iiiiiii.....iiiiiiiii.......iiiiiiiiiiiiiiiiiiiiiii........ 1600/2874
i...............iiii....................................................................i.i........i 1700/2874
....i......i..........................................i..........iiiiiiiiiii.iii......i.i...i....iii 1800/2874
ii..........................i...........i................i.............i..iiii...................... 1900/2874
.....i..ii...i.......i...............i..........iii....i.........................ii...ii...ii.i.i.i. 2000/2874
i.i.i.i.i.i.i.i.i.i.i.i.i.i.iiiiiiiii....................................iiiii...................... 2100/2874
.............................................................i.........iiiiiiiiiiiiiiii.iiiiiiii.i.. 2200/2874
i......iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii...ii.......i............ 2300/2874
...............i..................iiiiii..........iiiiiii.....iiiiiiiii.......iiiiiiiiiiiiiiiiiiiiii 2400/2874
i........i...............................................................i.i....i.i.i............... 2500/2874
.......................................................i.i....i.i.i................................. 2600/2874
.....................................i.i....i.i.i.................................................i. 2700/2874
.................................i..................................i...................test real_1_9_0::nnef::test_bvlc_alexnet has been running for over 60 seconds
test real_1_9_0::nnef::test_resnet50 has been running for over 60 seconds
test real_1_9_0::nnef::test_vgg19 has been running for over 60 seconds
.test real_1_9_0::nnef::test_zfnet512 has been running for over 60 seconds
test real_1_9_0::optim::test_bvlc_alexnet has been running for over 60 seconds
........test real_1_9_0::optim::test_vgg19 has been running for over 60 seconds
test real_1_9_0::optim::test_zfnet512 has been running for over 60 seconds
... 2800/2874
..iiiiiiiiii...iiiiii....iiiiiiiiii...iiiiii....iiiiiiiiii...iiiiii.......
test result: ok. 2050 passed; 0 failed; 824 ignored; 0 measured; 0 filtered out; finished in 174.71s


running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

+ ./.travis/cli-tests.sh
+ cargo -q build -q -p tract --release
+ ./.travis/cache_file.sh ARM-ML-KWS-CNN-M.pb GRU128KeywordSpotter-v2-10epochs.onnx hey_snips_v4_model17.pb hey_snips_v4_model17.alpha1.tgz inception_v3_2016_08_28_frozen.pb mdl-en-2019-Q3-librispeech.onnx mobilenet_v1_1.0_224_frozen.pb mobilenet_v2_1.4_224_frozen.pb squeezenet.onnx en_libri_real.tar.gz
+ '[' -d en_libri_real ']'
+ tar zxf en_libri_real.tar.gz
+ tract/.cached/squeezenet.onnx run -q --assert-output 1,1000,1,1,f32
[2021-07-23T14:04:58.574840000Z WARN  tract::tensor] Using random input for input called "data_0": 1,3,224,224,F32
+ tract/.cached/inception_v3_2016_08_28_frozen.pb -i 1,299,299,3,f32 run -q --assert-output-fact 1,1001,f32
[2021-07-23T14:04:59.957510000Z WARN  tract::tensor] Using random input for input called "input": 1,299,299,3,F32
+ tract/.cached/inception_v3_2016_08_28_frozen.pb -i 1,299,299,3,f32 -O run -q --assert-output-fact 1,1001,f32
[2021-07-23T14:05:02.004590000Z WARN  tract::tensor] Using random input for input called "input": 1,299,299,3,F32
+ tract/.cached/ARM-ML-KWS-CNN-M.pb -O -i 49,10,f32 --partial --input-node Mfcc run -q
[2021-07-23T14:05:02.305684000Z WARN  tract::tensor] Using random input for input called "Mfcc": 49,10,F32
+ tract/.cached/mdl-en-2019-Q3-librispeech.onnx -O -i S,40,f32 --output-node output --pulse 24 run -q
[2021-07-23T14:05:02.519637000Z WARN  tract::tensor] Using random input for input called "input": 24,40,F32
+ tract/.cached/mobilenet_v1_1.0_224_frozen.pb -O -i 1,224,224,3,f32 run -q --assert-output-fact 1,1001,f32
[2021-07-23T14:05:02.789422000Z WARN  tract::tensor] Using random input for input called "input": 1,224,224,3,F32
+ tract/.cached/mobilenet_v2_1.4_224_frozen.pb -O -i 1,224,224,3,f32 run -q --assert-output-fact 1,1001,f32
[2021-07-23T14:05:03.161343000Z WARN  tract::tensor] Using random input for input called "input": 1,224,224,3,F32
+ tract/.cached/GRU128KeywordSpotter-v2-10epochs.onnx -O run -q --assert-output-fact 1,3,f32
[2021-07-23T14:05:03.303131000Z WARN  tract::tensor] Using random input for input called "input.1": 1,1,80,F32
+ tract/.cached/hey_snips_v4_model17.pb -i S,20,f32 --pulse 8 dump --cost -q --assert-cost 'FMA(F32)=2060448,Div(F32)=24576,Buffer(F32)=2920,Params(F32)=222250'
Cost summary
 * Div(F32): 24576
 * FMA(F32): 2060448
 * Buffer(F32): 2920
 * Params(F32): 222250
+ tract/.cached/mdl-en-2019-Q3-librispeech.onnx -O -i S,40,f32 --output-node output --pulse 24 dump -q --assert-op-count Add 6 --assert-op-count Mul 22 --assert-op-count Max 0 --assert-op-count LirMatMulUnary 27 --assert-op-count MatMatMulPack 10
+ tract/.cached/hey_snips_v4_model17.pb -i S,20,f32 dump -q --assert-op-count AddAxis 0
+ tract/.cached/en_libri_real/model.raw.txt -f kaldi --output-node output --kaldi-downsample 3 --kaldi-left-context 5 --kaldi-right-context 15 --kaldi-adjust-final-offset -5 --input-bundle tract/.cached/en_libri_real/io.npz run --assert-output-bundle tract/.cached/en_libri_real/io.npz
+ tract/.cached/en_libri_real/model.raw -f kaldi --output-node output --kaldi-downsample 3 --kaldi-left-context 5 --kaldi-right-context 15 --kaldi-adjust-final-offset -5 --input-bundle tract/.cached/en_libri_real/io.npz run --assert-output-bundle tract/.cached/en_libri_real/io.npz
+ tract/.cached/en_libri_real/model.onnx --output-node output --kaldi-left-context 5 --kaldi-right-context 15 --kaldi-adjust-final-offset -5 --input-bundle tract/.cached/en_libri_real/io.npz run --assert-output-bundle tract/.cached/en_libri_real/io.npz
+ for t in 'harness/pre-optimized-graphes/*'
++ pwd
+ export TRACT_RUN=tract/target/release/tract
+ TRACT_RUN=tract/target/release/tract
+ cd harness/pre-optimized-graphes/hey_snips_v4_model17
+ ./runme.sh
+ '[' -z tract/.cached ']'
+ : tract/target/release/tract
+ tract/target/release/tract tract/.cached/hey_snips_v4_model17.pb -i S,20,f32 --pulse 8 --nnef-tract-pulse dump -q --nnef-graph found
+ diff expected found
+ for t in 'harness/pre-optimized-graphes/*'
++ pwd
+ export TRACT_RUN=tract/target/release/tract
+ TRACT_RUN=tract/target/release/tract
+ cd harness/pre-optimized-graphes/mdl-en-2019-Q3-librispeech
+ ./runme.sh
+ '[' -z tract/.cached ']'
+ : tract/target/release/tract
+ tract/target/release/tract tract/.cached/mdl-en-2019-Q3-librispeech.onnx -i S,40 --output-node output --pulse 24 --nnef-tract-pulse dump -q --nnef-graph found
+ diff expected found
+ '[' -e .aws/credentials ']'
+ cd kaldi/test_cases
+ TRACT_RUN=../../target/release/tract
+ ./run_all.sh
./delay_1_D10_T15_S1 (txt) OK
./delay_1_D10_T15_S1 (bin) OK
./delay_1_D10_T15_S1 (bin-opti) OK
./delay_3_D7_T17_S1 (txt) OK
./delay_3_D7_T17_S1 (bin) OK
./delay_3_D7_T17_S1 (bin-opti) OK
./delay_5_D15_T24_S1 (txt) OK
./delay_5_D15_T24_S1 (bin) OK
./delay_5_D15_T24_S1 (bin-opti) OK
./delay_5_D20_T259_S3 (txt) OK
./delay_5_D20_T259_S3 (bin) OK
./delay_5_D20_T259_S3 (bin-opti) OK
./delay_5_D5_T43_S3 (txt) OK
./delay_5_D5_T43_S3 (bin) OK
./delay_5_D5_T43_S3 (bin-opti) OK
./fixed_affine_40x10_T40_S3 (txt) OK
./fixed_affine_40x10_T40_S3 (bin) OK
./fixed_affine_40x10_T40_S3 (bin-opti) OK
./fixed_affine_40x64_T25_S1 (txt) OK
./fixed_affine_40x64_T25_S1 (bin) OK
./fixed_affine_40x64_T25_S1 (bin-opti) OK
./lstm_10-4-6_T10_S1_delay1 (txt) OK
./lstm_10-4-6_T10_S1_delay1 (bin) OK
./lstm_10-4-6_T10_S1_delay1 (bin-opti) OK
./lstm_12-3-6_T40_S1_delay2 (txt) OK
./lstm_12-3-6_T40_S1_delay2 (bin) OK
./lstm_12-3-6_T40_S1_delay2 (bin-opti) OK
./lstm_15-7-8_T30_S2_delay2 (txt) OK
./lstm_15-7-8_T30_S2_delay2 (bin) OK
./lstm_15-7-8_T30_S2_delay2 (bin-opti) OK
./lstm_5-8-16_T20_S3_delay1 (txt) OK
./lstm_5-8-16_T20_S3_delay1 (bin) OK
./lstm_5-8-16_T20_S3_delay1 (bin-opti) OK
./lstm_nonlin_10_T34_S1_delay3 (txt) OK
./lstm_nonlin_10_T34_S1_delay3 (bin) OK
./lstm_nonlin_10_T34_S1_delay3 (bin-opti) OK
./lstm_nonlin_15_T42_S3_delay3 (txt) OK
./lstm_nonlin_15_T42_S3_delay3 (bin) OK
./lstm_nonlin_15_T42_S3_delay3 (bin-opti) OK
./lstm_nonlin_7_T23_S1_delay1 (txt) OK
./lstm_nonlin_7_T23_S1_delay1 (bin) OK
./lstm_nonlin_7_T23_S1_delay1 (bin-opti) OK
./lstm_without_lstm_test1 (txt) OK
./lstm_without_lstm_test1 (bin) OK
./lstm_without_lstm_test1 (bin-opti) OK
./lstm_without_lstm_test2_delay1_S1 (txt) OK
./lstm_without_lstm_test2_delay1_S1 (bin) OK
./lstm_without_lstm_test2_delay1_S1 (bin-opti) OK
./lstm_without_lstm_test2_delay3_S1 (txt) OK
./lstm_without_lstm_test2_delay3_S1 (bin) OK
./lstm_without_lstm_test2_delay3_S1 (bin-opti) OK
./lstm_without_lstm_test2_delay3_S3 (txt) OK
./lstm_without_lstm_test2_delay3_S3 (bin) OK
./lstm_without_lstm_test2_delay3_S3 (bin-opti) OK
./ngaffine_10x2_T13_S2 (txt) OK
./ngaffine_10x2_T13_S2 (bin) OK
./ngaffine_10x2_T13_S2 (bin-opti) OK
./ngaffine_12x5_T20_S1_o-1_0_1 (txt) OK
./ngaffine_12x5_T20_S1_o-1_0_1 (bin) OK
./ngaffine_12x5_T20_S1_o-1_0_1 (bin-opti) OK
./ngaffine_12x5_T20_S1_o-1_0_1__identity (txt) OK
./ngaffine_12x5_T20_S1_o-1_0_1__identity (bin) OK
./ngaffine_12x5_T20_S1_o-1_0_1__identity (bin-opti) OK
./ngaffine_17x3_T29_S1_o-3_0_3 (txt) OK
./ngaffine_17x3_T29_S1_o-3_0_3 (bin) OK
./ngaffine_17x3_T29_S1_o-3_0_3 (bin-opti) OK
./ngaffine_23x15_T1_S1 (txt) OK
./ngaffine_23x15_T1_S1 (bin) OK
./ngaffine_23x15_T1_S1 (bin-opti) OK
./ngaffine_40x10_T15_S1 (txt) OK
./ngaffine_40x10_T15_S1 (bin) OK
./ngaffine_40x10_T15_S1 (bin-opti) OK
./relu_D10_T11 (txt) OK
./relu_D10_T11 (bin) OK
./relu_D10_T11 (bin-opti) OK
./relu_D1_T21 (txt) OK
./relu_D1_T21 (bin) OK
./relu_D1_T21 (bin-opti) OK
./relu_D23_T1 (txt) OK
./relu_D23_T1 (bin) OK
./relu_D23_T1 (bin-opti) OK
./renorm_tgt0.1_D17_T7 (txt) OK
./renorm_tgt0.1_D17_T7 (bin) OK
./renorm_tgt0.1_D17_T7 (bin-opti) OK
./renorm_tgt1.0_D20_T35 (txt) OK
./renorm_tgt1.0_D20_T35 (bin) OK
./renorm_tgt1.0_D20_T35 (bin-opti) OK
./renorm_tgt1.0_D24_T32_input0 (txt) OK
./renorm_tgt1.0_D24_T32_input0 (bin) OK
./renorm_tgt1.0_D24_T32_input0 (bin-opti) OK
./renorm_tgt1.5_D7_T5 (txt) OK
./renorm_tgt1.5_D7_T5 (bin) OK
./renorm_tgt1.5_D7_T5 (bin-opti) OK
+ cd onnx/test_cases
+ TRACT_RUN=../../target/release/tract
+ ./run_all.sh
./lgbm_classifier_tensor (plain) OK
./lgbm_classifier_tensor (decl) OK
./lgbm_classifier_tensor (opti) OK
./lgbm_classifier_tensor (nnef) OK
./lgbm_regressor_tensor (plain) ignored
./lgbm_regressor_tensor (decl) ignored
./lgbm_regressor_tensor (opti) ignored
./lgbm_regressor_tensor (nnef) ignored
./qlstm_3-2-3_T3_S1 (plain) OK
./qlstm_3-2-3_T3_S1 (decl) OK
./qlstm_3-2-3_T3_S1 (opti) OK
./qlstm_3-2-3_T3_S1 (nnef) ignored
./qrelu_1 (plain) OK
./qrelu_1 (decl) OK
./qrelu_1 (opti) OK
./qrelu_1 (nnef) ignored
./qrelu_2 (plain) OK
./qrelu_2 (decl) OK
./qrelu_2 (opti) OK
./qrelu_2 (nnef) ignored
./qsigmoid_1 (plain) OK
./qsigmoid_1 (decl) OK
./qsigmoid_1 (opti) OK
./qsigmoid_1 (nnef) ignored
./qsigmoid_2 (plain) OK
./qsigmoid_2 (decl) OK
./qsigmoid_2 (opti) OK
./qsigmoid_2 (nnef) ignored
./qtanh_1 (plain) OK
./qtanh_1 (decl) OK
./qtanh_1 (opti) OK
./qtanh_1 (nnef) ignored
./qtanh_2 (plain) OK
./qtanh_2 (decl) OK
./qtanh_2 (opti) OK
./qtanh_2 (nnef) ignored
./qtdnn_10x5_101_i32_biases (plain) OK
./qtdnn_10x5_101_i32_biases (decl) OK
./qtdnn_10x5_101_i32_biases (opti) OK
./qtdnn_10x5_101_i32_biases (nnef) ignored
./transformer-mlm (plain) OK
./transformer-mlm (decl) OK
./transformer-mlm (opti) OK
./transformer-mlm (nnef) ignored
./xgboost_classifier_tree (plain) OK
./xgboost_classifier_tree (decl) OK
./xgboost_classifier_tree (opti) OK
./xgboost_classifier_tree (nnef) OK
./xgboost_regressor_tree (plain) ignored
./xgboost_regressor_tree (decl) ignored
./xgboost_regressor_tree (opti) ignored
./xgboost_regressor_tree (nnef) ignored
</details>
